### PR TITLE
Adding API functionality and better usability

### PR DESCRIPTION
**Pull request by Dzari**

### CODE REVIEW

**Review by gennady-bars — COMMENTED** [review]
Dear student, (**Please, expand the general comment ↓**)

Unfortunately I have to reject the project without reviewing:


* LIkes counter doesn't change https://yadi.sk/d/SyisDUD3M_6RSg
* When I like new cards there are errors https://yadi.sk/d/U3hlOospJHJhYA 
* When I delete new cards there are errors https://yadi.sk/d/sYfKCX8YthcTiw 
* When I change the avatar and reload the page the old avatar appears again

**Comment by Dzari** [comment]
I wasn't able to recreate the new card bug errors but I fixed the like counter and the delete card errors. The avatar is now set via the API on load rather than HTML.

**Review by gennady-bars — COMMENTED** [review]
Dear student, (**Please, expand the general comment ↓**)

Unfortunately I have to reject the project without reviewing:


* When I like **new** cards there are errors https://yadi.sk/d/U3hlOospJHJhYA . How to check it: add 3 cards in a row and try to like them
* When I delete **new** cards and reload the page they all appear again
* I can't delete **old** cards. See the errors https://yadi.sk/d/lw0m5wHmobFrwQ

**Comment by Dzari** [comment]
Finally got it working. Should be ready to review - probably going to need a lot of cleaning up/refactoring

File: src/components/Api.js
Comment (gennady-bars) [inline src/components/Api.js]: It’s better not to use `catch` in `Api` because you need to `catch` possible errors at the very **end** of any request and this end is in `index.js`   after all blocks `then`. And in real projects blocks `catch` usually open error popups and so on.

File: src/components/Api.js
Comment (gennady-bars) [inline src/components/Api.js]: **Could be improved**

You can make a special method (function ) for fetching and checking responses **not to duplicate** it in every request:

```js

  _request(url, options) {
    return fetch(url, options).then(this.validateAndParse)
  }

```

And now you can safely replace `fetch` with `this._request` and remove repeated checking. And you don’t need to change anything else.

File: src/components/Card.js
Comment (gennady-bars) [inline src/components/Card.js]: You need to **toggle likes** only in block `then` after a successful response from the server because if there is an error the user will not know about that and will see the toggled heart color.

File: src/components/Card.js
Comment (gennady-bars) [inline src/components/Card.js]: You need to **toggle likes** only in block `then` after a successful response from the server because if there is an error the user will not know about that and will see the toggled heart color.

File: src/components/Card.js
Comment (gennady-bars) [inline src/components/Card.js]: You need to **toggle likes** only in block `then` after a successful response from the server because if there is an error the user will not know about that and will see the toggled heart color.

File: src/pages/index.js
Comment (gennady-bars) [inline src/pages/index.js]: You need to `catch` possible errors at the very **end** of any server request.

**Please, correct it everywhere**

File: src/pages/index.js
Comment (gennady-bars) [inline src/pages/index.js]: Class `UserInfo` should set the entire user’s data (including the **avatar**)

File: src/pages/index.js
Comment (gennady-bars) [inline src/pages/index.js:115]: `.renderLikes();` should be inside `Card` not to duplicate it in different places

File: src/pages/index.js
Comment (gennady-bars) [inline src/pages/index.js]: `api.getUserInfo();` is not needed for avatar updating because you need to use the `response` from the server when you call it in line 40

File: src/pages/index.js
Comment (gennady-bars) [inline src/pages/index.js:198]: **Could be improved**

It’s better to make method `renderLoading` inside class `PopupWithForm`. You should find the button once in the constructor and change the text:  

```js
  constructor({ popupSelector, handleSubmit }) {
    super(popupSelector);
    // fix the initial button text only once in the constructor
    this._submitBtnText = this._submitBtn.textContent
  }
  // add 2 params: isLoading and loadingText with a default text
  renderLoading(isLoading, loadingText='Saving...') {
    if (isLoading) {
      this._submitBtn.textContent = loadingText;
    } else {
 // here we return back the initial text. So, you don’t need to bother yourself about it
      this._submitBtn.textContent = this._submitBtnText;
    }
  }

```

File: src/pages/index.js
Comment (gennady-bars) [inline src/pages/index.js]: You need to update `DOM` only in blocks `then` after a successful response from the server so that when there is an error it wouldn’t change

File: src/pages/index.js
Comment (gennady-bars) [inline src/pages/index.js]: This should be in block `then`

You need to **reset** the form only  in block `then` not to do that immediately after `submit` because when there is an error the popup will still be opened and the user can try to send the data again to the server without having to type the data into the inputs once more

File: src/pages/index.js
Comment (gennady-bars) [inline src/pages/index.js]: You need to return the **default** button text only in blocks `finally` so that the button could get the text back in any case: `success` or `failure`.

File: src/pages/index.js
Comment (gennady-bars) [inline src/pages/index.js]: You need to **delete** cards only in block `then` after a **successful** response from the server so that if there is an error the card would stay in the `DOM`.

File: src/pages/index.js
Comment (gennady-bars) [inline src/pages/index.js:269]: **Could be improved**


If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}

```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```

File: src/pages/index.js
Comment (gennady-bars) [inline src/pages/index.js]: **Could be improved**

You can **universally** create instances of validators for all forms in the project storing them inside one object `formValidators`.  And then you can take any validator using attribute `name` of the form where you need to disable/enable the submit button or remove errors.

```js
// define an object for storing validators
const formValidators = {}

const enableValidation = (config) => {
  const formList = Array.from(document.querySelectorAll(config.formSelector))
  formList.forEach((formElement) => {
    const validator = new FormValidator(config, formElement)
    // here you get the name of the form (if you don’t have it then you need to add it into each form in `index.html` first)
    const formName = formElement.getAttribute('name')

   // here you store the validator using the `name` of the form
    formValidators[formName] = validator;
    validator.enableValidation();
  });
};

enableValidation(config);

```

And now you can use them for disabling buttons or clearing errors:

```js

formValidators[ profileForm.getAttribute('name') ].resetValidation()

// or you can use a string – the name of the form (you know it from `index.html`)

formValidators['profile-form'].resetValidation()

```

**Review by gennady-bars — COMMENTED** [review]
Dear student, (**Please, expand the general comment ↓**)

Good job but there are some remarks:



* I delete a card from the middle of the card list but only the 1rst card in the row is deleted.
* According to the task you need to change submit buttons text (`Saving…`) while waiting for the server response. And the user should see it. 
* It’s better not to use `catch` in `Api` because you need to `catch` possible errors at the very **end** of any request and this end is in `index.js`   after all blocks `then`. And in real projects blocks `catch` usually open error popups and so on.
* You need to **toggle likes** only in block `then` after a successful response from the server because if there is an error the user will not know about that and will see the toggled heart color.
* You need to `catch` possible errors at the very **end** of any server request.
* Class `UserInfo` should set the entire user’s data (including the **avatar**)
* You need to update `DOM` only in blocks `then` after a successful response from the server so that when there is an error it wouldn’t change
* You need to **reset** the form only  in block `then` not to do that immediately after `submit` because when there is an error the popup will still be opened and the user can try to send the data again to the server without having to type the data into the inputs once more
* You need to close popups only in blocks `then` after a **successful response** from the server so that when there is an error it would be opened (the inputs will not be cleared and the user could send the data again to the server) and if you change the button text you could see the change as well
* You need to return the **default** button text only in blocks `finally` so that the button could get the text back in any case: `success` or `failure`.
* You need to **delete** cards only in block `then` after a **successful** response from the server so that if there is an error the card would stay in the `DOM`.

**Review by gennady-bars — COMMENTED** [review]
Dear student, (**Please, expand the general comment ↓**)

Unfortunately I have to reject the project without reviewing:



* I can't delete cards. See the errors https://yadi.sk/d/yg737CUXTCWhRg

**Review by gennady-bars — COMMENTED** [review]
Dear student, (**Please, expand the general comment ↓**)

Unfortunately I have to reject the project without reviewing because you haven't corrected some of my remarks



* According to the task you need to change submit buttons text (`Saving…`) while waiting for the server response. And the user should see it.
* You haven't corrected my remark https://github.com/Dzari/se_project_aroundtheus/pull/1#discussion_r1630195089 
* See all my remarks in the code starting from here  https://github.com/Dzari/se_project_aroundtheus/pull/1#pullrequestreview-2103149695

**Comment by Dzari** [comment]
I have removed the catches from the API but as for changing the save button to saving... when waiting for the server response - I am already doing that. If you add a breakpoint at any fetch request you can see it change to saving...

**Comment by Dzari** [comment]
got some clarification on the saving... issue. Should show up when users click save/create

File: src/pages/index.js
Comment (gennady-bars) [inline src/pages/index.js]: Please, delete `setProfilePic()` because you call the same `api.getUserInfo()` inside `setUserInfo` in line 151, so you need to set the avatar there as well

File: src/pages/index.js
Comment (gennady-bars) [inline src/pages/index.js]: Class `UserInfo` should set the entire user’s data (including the **avatar**)

File: src/pages/index.js
Comment (gennady-bars) [inline src/pages/index.js]: Class `UserInfo` should set the entire user’s data (including the **avatar**)

File: src/pages/index.js
Comment (gennady-bars) [inline src/pages/index.js]: Please, add `await` to wait for the server response otherwise the card is deleted immediately

File: src/pages/index.js
Comment (gennady-bars) [inline src/pages/index.js]: You need to close popups only in blocks `then` after a **successful response** from the server so that when there is an error it would be opened (the inputs will not be cleared and the user could send the data again to the server) and if you change the button text you could see the change as well

File: src/pages/index.js
Comment (gennady-bars) [inline src/pages/index.js]: Please, don't use `try/catch` if you don't know how to handle server responses. You need to wait for the server requests

It's better to use `then...catch...finally` rather than `try/catch` to make it easier for you

File: src/pages/index.js
Comment (gennady-bars) [inline src/pages/index.js]: This should be in block `then`

You need to update `DOM` only in blocks `then` after a successful response from the server so that when there is an error it wouldn’t change

File: src/pages/index.js
Comment (gennady-bars) [inline src/pages/index.js]: * Class `UserInfo` should set the entire user’s data (including the **avatar**)

File: src/pages/index.js
Comment (gennady-bars) [inline src/pages/index.js]: * You need to close popups only in blocks `then` after a **successful response** from the server so that when there is an error it would be opened (the inputs will not be cleared and the user could send the data again to the server) and if you change the button text you could see the change as well

File: src/pages/index.js
Comment (gennady-bars) [inline src/pages/index.js]: Now only class `Card` knows every **card selector**, so you need to encapsulate the likes logics inside its methods and call them in `index.js`. It’s a violation of **Single responsibility** principle

**Review by gennady-bars — COMMENTED** [review]
One more step and everything will be fine:


* Please, delete `setProfilePic()` because you call the same `api.getUserInfo()` inside `setUserInfo` in line 151, so you need to set the avatar there as well 
* Class `UserInfo` should set the entire user’s data (including the **avatar**)
* You need to close popups only in blocks `then` after a **successful response** from the server so that when there is an error it would be opened (the inputs will not be cleared and the user could send the data again to the server) and if you change the button text you could see the change as well
* Please, don't use `try/catch` if you don't know how to handle server responses. You need to wait for the server requests
* You need to update `DOM` only in blocks `then` after a successful response from the server so that when there is an error it wouldn’t change
* Now only class `Card` knows every **card selector**, so you need to encapsulate the likes logics inside its methods and call them in `index.js`. It’s a violation of **Single responsibility** principle

**Comment by Dzari** [comment]
Everything is in .then blocks as they should be, the avatar is updated with the other userinfo, and I updated the logic for liking. On page load the default value is data.isliked but when user likes a card a true/false value is sent to card.renderlikes so the card elements are all in the card class. I will add a helper function to the submit functions to change the submit button text after approval is complete

File: src/pages/index.js
Comment (gennady-bars) [inline src/pages/index.js]: Class `UserInfo` should set the entire user’s data (including the **avatar**)

You should use `userInfo.setUserInfo` here


```js

    .then((userData) => {
      userInfo.setUserInfo(userData)
    })
```

File: src/pages/index.js
Comment (gennady-bars) [inline src/pages/index.js]: Class `UserInfo` should set the entire user’s data (including the **avatar**)

You should use `userInfo.setUserInfo` here


```js

    .then((userData) => {
      userInfo.setUserInfo(userData)
    })
```

File: src/pages/index.js
Comment (gennady-bars) [inline src/pages/index.js]: `_data`

Please, delete the low dash (`underscore`) from methods if you want to use them outside a class because the low dash shows a private method which is used only inside the class.

File: src/pages/index.js
Comment (gennady-bars) [inline src/pages/index.js]: `_data`

Please, delete the low dash (`underscore`) from methods if you want to use them outside a class because the low dash shows a private method which is used only inside the class.

**Review by gennady-bars — COMMENTED** [review]
One more step is needed:


* Class `UserInfo` should set the entire user’s data (including the **avatar**)
* Please, delete the low dash (`underscore`) from methods if you want to use them outside a class because the low dash shows a private method which is used only inside the class.

**Review by gennady-bars — APPROVED** [review]
You’ve done a great job.

Your project has been accepted

Good luck

**Pull request merged**
